napari/_qt/menus/_util.py

```diff
@@ -0,0 +1,103 @@
+from typing import TYPE_CHECKING, Callable, List, Union
+
+from qtpy.QtWidgets import QAction, QMenu
+
+if TYPE_CHECKING:
+    from typing_extensions import TypedDict
+
+    from ...utils.events import EventEmitter
+
+    try:
+        from qtpy.QtCore import SignalInstance
+    except ImportError:
+        from qtpy.QtCore import pyqtBoundSignal as SignalInstance
+
+    class ActionDict(TypedDict):
+        text: str
+        # these are optional
+        slot: Callable
+        shortcut: str
+        statusTip: str
+        menuRole: QAction.MenuRole
+        checkable: bool
+        checked: bool
+        check_on: Union[EventEmitter, SignalInstance]
+
+    class MenuDict(TypedDict):
+        menu: str
+        # these are optional
+        items: List[ActionDict]
+
+    # note: TypedDict still doesn't have the concept of "optional keys"
+    # so we add in generic `dict` for type checking.
+    # see PEP655: https://www.python.org/dev/peps/pep-0655/
+    MenuItem = Union[MenuDict, ActionDict, dict]
+
+
+def populate_menu(menu: QMenu, actions: List['MenuItem']):
```

Comment:
How will this work for more dynamic changes, that do not require recreating a menu everytime? or is that the pattern we want, to always recreate the menu from scratch?

Reply:
it's just the pattern we have at the moment... (i.e. this is how these menus were already being constructed and used).  However, if we find we need something more dynamic in the future, it's still possible to get a handle on internal menus and modify them.  we can add some conveniences around that when we have a use case

Reply:
👍🏼